storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java

```diff
@@ -253,125 +256,139 @@ private void startConsumerThread(final KafkaConsumer<String, Str.StorageValue> c
-    public void updateArtifactState(String artifactId, ArtifactState state, Integer version) {
-        ArtifactVersionMetaDataDto metadata = sqlStorage.getArtifactVersionMetaData(artifactId, version);
-        updateArtifactState(metadata.getState(), artifactId, version, state);
-    }
-
+    /**
+     * @see io.apicurio.registry.storage.RegistryStorage#createArtifact(java.lang.String, io.apicurio.registry.types.ArtifactType, io.apicurio.registry.content.ContentHandle)
+     */
     @Override
     public CompletionStage<ArtifactMetaDataDto> createArtifact(String artifactId, ArtifactType artifactType, ContentHandle content) throws ArtifactAlreadyExistsException, RegistryStorageException {
-
-        try {
-            sqlStorage.getArtifactMetaData(artifactId);
-            throw new ArtifactAlreadyExistsException(artifactId);
-        } catch (ArtifactNotFoundException e) {
-            //ignored
-            //artifact does not exist, we can create it
-        }
-
-        return submitter
-                .submitArtifact(Str.ActionType.CREATE, artifactId, -1, artifactType, content.bytes())
-                .thenCompose(reqId -> {
-                    return (CompletionStage<ArtifactMetaDataDto>) coordinator.waitForResponse(reqId);
-                });
+        return createArtifactWithMetadata(artifactId, artifactType, content, null);
     }
 
+    /**
+     * @see io.apicurio.registry.storage.RegistryStorage#createArtifactWithMetadata(java.lang.String, io.apicurio.registry.types.ArtifactType, io.apicurio.registry.content.ContentHandle, io.apicurio.registry.storage.EditableArtifactMetaDataDto)
+     */
     @Override
     public CompletionStage<ArtifactMetaDataDto> createArtifactWithMetadata(String artifactId, ArtifactType artifactType,
             ContentHandle content, EditableArtifactMetaDataDto metaData) throws ArtifactAlreadyExistsException, RegistryStorageException {
-        return createArtifact(artifactId, artifactType, content)
-            .thenCompose(amdd ->
-                submitter.submitMetadata(Str.ActionType.UPDATE, artifactId, -1, metaData.getName(), metaData.getDescription(), metaData.getLabels(), metaData.getProperties())
-                    .thenAccept(reqId -> coordinator.waitForResponse(reqId))
-                    .thenApply(v -> DtoUtil.setEditableMetaDataInArtifact(amdd, metaData))
-            );
-
+        if (sqlStore.isArtifactExists(artifactId)) {
+            throw new ArtifactAlreadyExistsException(artifactId);
+        }
+        
+        String contentHash = ensureContent(content, artifactId, artifactType);
+        String createdBy = null; // TODO populate when authentication is available
+        Date createdOn = new Date();
+        
+        return submitter
+                .submitArtifact(artifactId, ActionType.Create, artifactType, contentHash, createdBy, createdOn, metaData)
+                .thenCompose(reqId -> (CompletionStage<ArtifactMetaDataDto>) coordinator.waitForResponse(reqId));
```

Comment:
here we would catch if there is any error, and we would use log compaction to remove the specific kafka message for this action (using it's uuid)

Reply:
Yes I will create a GH issue for this.